patches/api/0382-More-Teleport-API.patch

```diff
@@ -0,0 +1,282 @@
++package io.papermc.paper.entity;
++
++import org.bukkit.Location;
++import org.bukkit.event.player.PlayerTeleportEvent;
++
++/**
++ * Represents coordinates in a teleportation that should be handled relatively.
++ *
++ * @see org.bukkit.entity.Player#teleport(Location, PlayerTeleportEvent.TeleportCause, boolean, boolean, RelativeTeleportFlag...)
++ */
++public enum RelativeTeleportFlag {
++    /**
++     * Represents the player's X coordinate
++     */
++    X,
++    /**
++     * Represents the player's Y coordinate
++     */
++    Y,
++    /**
++     * Represents the player's Z coordinate
++     */
++    Z,
++    /**
++     * Represents the player's yaw
++     */
++    YAW,
++    /**
++     * Represents the player's pitch
++     */
++    PITCH;
++
++}
+diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
+index 8bc6876c82935988436597161fa0ec94c032174b..6696b5965aada0b31c123e4744448e8661617d9f 100644
+--- a/src/main/java/org/bukkit/entity/Entity.java
++++ b/src/main/java/org/bukkit/entity/Entity.java
+@@ -121,10 +121,73 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
+      *
+      * @param yaw the yaw
+      * @param pitch the pitch
+-     * @throws UnsupportedOperationException if used for players
+      */
+     public void setRotation(float yaw, float pitch);
+ 
++    // Paper start - Teleport API
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
++     * will cause this teleportation to return false.
++     *
++     * @param location New location to teleport this entity to
++     * @param ignorePassengers If passengers should be ignored when teleporting
++     * @return <code>true</code> if the teleport was successful
++     */
++    default boolean teleport(@NotNull Location location, boolean ignorePassengers) {
++        return teleport(location, TeleportCause.PLUGIN, ignorePassengers);
++    }
++
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
++     * will cause this teleportation to return false.
++     *
++     * @param location New location to teleport this entity to
++     * @param cause The cause of this teleportation
++     * @param ignorePassengers If passengers should be ignored when teleporting
++     * @return <code>true</code> if the teleport was successful
++     */
++    default boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean ignorePassengers) {
++        return teleport(location, cause, ignorePassengers, true);
++    }
++
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: Teleporting to a different world with ignorePassengers to true while the entity has entities riding it
```

Comment:
I mean stuff like this, "return false" just sounds horrible and doesn't really describe that the teleportation will not occur

Reply:
Ahh I see, will do

Reply:
I've added "and not occur". I'd appreciate feedback on this.